Topology hints

bart0sh opened the pull request:
Implemented passing topology information from device plugins to kubelet in ListAndWatch responses.

topology.go and topology_test.go were taken from
https://github.com/intel/cri-resource-manager/blob/v0.2.0/pkg/sysfs/ with little changes on top of them.

rojkov commented on pkg/deviceplugin/topology.go:
```diff
@@ -0,0 +1,254 @@
+// MergeTopologyHints combines org and hints.
+func MergeTopologyHints(org, hints TopologyHints) (res TopologyHints) {
+	if org != nil {
+		res = org
```
Bear in mind that in certain circumstances this public function produces side effects by mutating its input (the `org` parameter). I guess this should be either documented or reworked.

rojkov commented on pkg/deviceplugin/topology.go:
```diff
@@ -0,0 +1,254 @@
+	return
+}
+
+// String returns the hints as a string.
```
Could be rephrased as
```
// String implements Stringer interface by returning hints as a string.
```

rojkov commented on pkg/deviceplugin/topology.go:
```diff
@@ -0,0 +1,254 @@
+// limitations under the License.
+
+// Most of this code was copied from CRI Resource Manager:
+// https://github.com/intel/cri-resource-manager/blob/v0.2.0/pkg/sysfs/topology.go
```
Wouldn't it make sense to just import the code?

mythi replied:
+1

I have a patch to deal with `vfio` devices. It'd be easier to submit it to cri-resource-manager and have it pulled here.

bart0sh replied:
go get fails to import it:
```
$ go get github.com/intel/cri-resource-manager/pkg/sysfs
go: finding github.com/intel/cri-resource-manager v0.2.0
go: finding github.com/intel/cri-resource-manager/pkg latest
go: finding github.com/intel/cri-resource-manager/pkg/sysfs latest
go: finding github.com/intel/cri-resource-manager v0.2.0
go: downloading github.com/intel/cri-resource-manager v0.2.0
verifying github.com/intel/cri-resource-manager@v0.2.0: github.com/intel/cri-resource-manager@v0.2.0: reading https://sum.golang.org/lookup/github.com/intel/cri-resource-manager@v0.2.0: 410 Gone
```

The actual error is this:
```
not found: unzip /tmp/gopath/pkg/mod/cache/download/github.com/intel/cri-resource-manager/@v/v0.2.0.zip: malformed file path "pkg/sysfs/testdata/sys/devices/pci0000:00/0000:00:02.0/class": invalid char ':'
```

I'll try to fix it in cri-rm.

bart0sh replied:
@mythi @rojkov here is the fix for ```go get``` issue: https://github.com/intel/cri-resource-manager/pull/154

There is also workaround for this issue. Either of two will hopefully merged soon.

bart0sh replied:
@mythi @rojkov here is a workaround I've mentioned before: https://github.com/intel/cri-resource-manager/pull/155

please review to speed up resolving this blocking issue.

bart0sh replied:
Importing even one module from cri-resource-manager creates a lot of entries in go.mod and go.sum even with [topology.go moved out of sysfs module](https://github.com/klihub/cri-resource-manager/tree/devel/split-out-topology/pkg/topology) e.g.
```
--- a/go.sum
+++ b/go.sum
@@ -2,6 +2,8 @@ bitbucket.org/bertimus9/systemstat v0.0.0-20180207000608-0eeff89b0690/go.mod h1:
 cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
 cloud.google.com/go v0.34.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
 cloud.google.com/go v0.38.0/go.mod h1:990N+gfupTy94rShfmMCWGDn0LpTmnzTp2qbd1dvSRU=
+contrib.go.opencensus.io/exporter/jaeger v0.1.0/go.mod h1:VYianECmuFPwU37O699Vc1GOcy+y8kOsfaxHRImmjbA=
+contrib.go.opencensus.io/exporter/prometheus v0.1.1-0.20191218042359-6151c48ac7fa/go.mod h1:dZQATH/0adcAA6FxxVxmOPuJhCsikLzs44Fv2B0V4ug=
 github.com/Azure/azure-sdk-for-go v21.4.0+incompatible/go.mod h1:9XXNKU+eRnpl9moKnB4QOLf1HestfXbmab5FXxiDBjc=
 github.com/Azure/azure-sdk-for-go v35.0.0+incompatible/go.mod h1:9XXNKU+eRnpl9moKnB4QOLf1HestfXbmab5FXxiDBjc=
 github.com/Azure/go-ansiterm v0.0.0-20170929234023-d6e3b3328b78/go.mod h1:LmzpDX56iTiv29bbRTIsUNlaFfuhWRQBWjQdVyAevI8=
@@ -36,12 +38,17 @@ github.com/PuerkitoBio/urlesc v0.0.0-20160726150825-5bd2802263f2/go.mod h1:uGdko
 github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578 h1:d+Bc7a5rLufV/sSk/8dngufqelfh6jnri85riMAaF/M=
 github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578/go.mod h1:uGdkoq3SwY9Y+13GIhn11/XLaGBb4BfwItxLd5jeuXE=
 github.com/Rican7/retry v0.1.0/go.mod h1:FgOROf8P5bebcC1DS0PdOQiqGUridaZvikzUmkFW6gg=
+github.com/Shopify/sarama v1.19.0/go.mod h1:FVkBWblsNy7DGZRfXLU0O9RCGt5g3g3yEuWXgklEdEo=
+github.com/Shopify/toxiproxy v2.1.4+incompatible/go.mod h1:OXgGpZ6Cli1/URJOF1DMxUHB2q5Ap20/P/eIdh4G0pI=
 github.com/StackExchange/wmi v0.0.0-20180116203802-5d049714c4a6/go.mod h1:3eOhrUMpNV+6aFIbp5/iudMxNCF27Vw2OZgy4xEx0Fg=
 github.com/agnivade/levenshtein v1.0.1/go.mod h1:CURSv5d9Uaml+FovSIICkLbAUZ9S4RqaHDIsdSBg7lM=
 github.com/alecthomas/template v0.0.0-20160405071501-a0175ee3bccc/go.mod h1:LOuyumcjzFXgccqObfd/Ljyb9UuFJ6TxHnclSeseNhc=
+github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751/go.mod h1:LOuyumcjzFXgccqObfd/Ljyb9UuFJ6TxHnclSeseNhc=
 github.com/alecthomas/units v0.0.0-20151022065526-2efee857e7cf/go.mod h1:ybxpYRFXyAe+OPACYpWeL0wqObRcbAqCMya13uyzqw0=
+github.com/alecthomas/units v0.0.0-20190717042225-c3de453c63f4/go.mod h1:ybxpYRFXyAe+OPACYpWeL0wqObRcbAqCMya13uyzqw0=
 github.com/andreyvit/diff v0.0.0-20170406064948-c7f18ee00883/go.mod h1:rCTlJbsFo29Kk6CurOXKm700vrz8f0KW0JNfpkRJY/8=
 github.com/anmitsu/go-shlex v0.0.0-20161002113705-648efa622239/go.mod h1:2FmKhYUyUczH0OGQWaF5ceTx0UBShxjsH6f8oGKYe2c=
+github.com/apache/thrift v0.12.0/go.mod h1:cp2SuWMxlEZw2r+iP2GNCdIi4C1qmUzdZFSVb+bacwQ=
 github.com/armon/circbuf v0.0.0-20150827004946-bbbad097214e/go.mod h1:3U/XgcO3hCbHZ8TKRvWD2dDTCfh9M9ya+I9JpbB7O8o=
 github.com/armon/consul-api v0.0.0-20180202201655-eb2c6b5be1b6/go.mod h1:grANhF5doyWs3UAsr3K4I6qtAmlQcZDesFNEHPZAzj8=
...
```

I think it's better to just copy the module or rewrite it as we don't need most of its functionality here.

mythi replied:
> I think it's better to just copy the module or rewrite it as we don't need most of its functionality here.

Yeah so we're back to this PR as it was originally. Can you take my VFIO patch in v2 for this PR?

bart0sh replied:
We found a solution to import topology APIs without polluting go.mod and go.sum files. It requires [this PR](https://github.com/intel/cri-resource-manager/pull/158) to be merged.

kad commented on pkg/deviceplugin/topology.go:
```diff
@@ -0,0 +1,254 @@
+// GetTopologyInfo returns topology information for the list of device nodes
+func GetTopologyInfo(devs []string) (*pluginapi.TopologyInfo, error) {
+	var result pluginapi.TopologyInfo
+	nodeIDs := map[int64]bool{}
```
it become more ugly in the assignments (like `struct{}{}`), there are other disadvantages that can be seen here: 
https://dave.cheney.net/2014/03/25/the-empty-struct#comment-2826

rojkov replied:
I wouldn't trust boolean-based sets because for example `len(m2)` can be misleading. In the long run sets implemented with empty structs are less error prone.

kad commented on pkg/deviceplugin/topology.go:
```diff
@@ -0,0 +1,254 @@
+			parentHints, er := NewTopologyHints(filepath.Dir(p))
+			if er == nil {
+				cpulist := map[string]bool{}
+				numalist := map[string]bool{}
```
struct{} is still a pointer to the struct header with empty fields, which is 64 bit at least.

rojkov replied:
&struct{}{} is a pointer, struct{}{} isn't.

From the blog post you mentioned:
```
type S struct {
        A struct{}
        B struct{}
}
var s S
fmt.Println(unsafe.Sizeof(s)) // prints 0
```
and
```
var x [1000000000]struct{}
fmt.Println(unsafe.Sizeof(x)) // prints 0
```

kad commented on pkg/deviceplugin/topology.go:
```diff
@@ -0,0 +1,254 @@
+		}
+		res[k] = v
+	}
+	return
```
in modern Go recommendations for short functions it is actually opposite. If the function fits into one screen (~10-15 lines), it is better to use naked returns to improve readability (less polluted).

rojkov replied:
I checked https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters 
> Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values. Corollary: it's not worth it to name result parameters just because it enables you to use naked returns. Clarity of docs is always more important than saving a line or two in your function.

I wouldn't interpret it as a recommendation for naked returns.

mythi commented on pkg/deviceplugin/api.go:
```diff
@@ -20,10 +20,32 @@ import (
+	}
+
+	topologyInfo, err := GetTopologyInfo(devPaths)
+	if err != nil {
```
make it `err == nil` :-)

bart0sh commented:
Imported topology API from cri-rm instead of copying it here. Addressed review comments.
Note that this PR is still WIP until [cri-rm PR](https://github.com/intel/cri-resource-manager/pull/158) is merged.

Please review again.

mythi commented:
> Please review again.

LGTM. Let's prepare the final import after intel/cri-resource-manager/pull/158 and intel/cri-resource-manager/pull/153

bart0sh commented:
split PR to two commits

bart0sh commented:
This PR is unblocked as both blocking PRs intel/cri-resource-manager/pull/158 and intel/cri-resource-manager/pull/153 have been merged.

Updated go.mod and go.sum to point to the cri-rm master.

@rojkov @mythi please review and merge.

mythi commented:
tested with `qat_plugin` and looks OK

```
DEBUG server.go:90 Sending to kubelet [&Device{ID:3f:01.1,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:01.2,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:01.4,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:01.6,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:01.7,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:02.1,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:da:01.0,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:01.3,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:01.7,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:02.4,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:02.5,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:3d:02.6,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:da:01.2,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:01.5,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:02.2,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:3d:01.1,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:01.3,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:01.4,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:02.0,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:02.3,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:02.0,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:01.0,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:02.1,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:02.7,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:da:02.0,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:3d:01.2,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:01.7,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:02.7,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:02.2,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:02.3,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:02.6,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:da:01.1,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:01.4,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:02.1,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:02.3,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:3d:01.5,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:01.6,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:02.5,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:01.5,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:02.4,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:02.5,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:01.3,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:02.2,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3d:02.4,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:3f:01.0,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:0,},},},} &Device{ID:da:01.6,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:02.6,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},} &Device{ID:da:02.7,Health:Healthy,Topology:&TopologyInfo{Nodes:[]*NUMANode{&NUMANode{ID:3,},},},}]
```

bart0sh commented:
Latest change: Replaced map[string]bool with map[string]struct{}{} as @rojkov  suggested in his review.
Please, review & merge.
